cli/build_run_result_registerer.go

```diff
@@ -0,0 +1,139 @@
+	stepStartTime time.Time,
+	step stepmanModels.StepModel,
+	stepInfoPtr stepmanModels.StepInfoModel,
+	stepIdxPtr int,
+	runIf string,
+	resultCode int,
+	exitCode int,
+	err error,
+	isLastStep bool,
+	redactedStepInputs map[string]string,
+	properties coreanalytics.Properties) {
+
+	timeout, noOutputTimeout := time.Duration(-1), time.Duration(-1)
+	if resultCode == models.StepRunStatusCodeFailed {
+		// Forward the status of a Step or a wrapped bitrise process.
+		switch exitCode {
+		case exitcode.CLIAbortedWithCustomTimeout:
+			resultCode = models.StepRunStatusAbortedWithCustomTimeout
+		case exitcode.CLIAbortedWithNoOutputTimeout:
+			resultCode = models.StepRunStatusAbortedWithNoOutputTimeout
+		}
+
+		var timeoutErr timeoutcmd.TimeoutError
+		if ok := errors.As(err, &timeoutErr); ok {
+			resultCode = models.StepRunStatusAbortedWithCustomTimeout
+			timeout = timeoutErr.Timeout
+		}
+
+		var noOutputTimeoutErr timeoutcmd.NoOutputTimeoutError
+		if ok := errors.As(err, &noOutputTimeoutErr); ok {
+			resultCode = models.StepRunStatusAbortedWithNoOutputTimeout
+			noOutputTimeout = noOutputTimeoutErr.Timeout
+		}
+	}
+
+	stepInfoCopy := stepmanModels.StepInfoModel{
+		Library:         stepInfoPtr.Library,
+		ID:              stepInfoPtr.ID,
+		Version:         stepInfoPtr.Version,
+		OriginalVersion: stepInfoPtr.OriginalVersion,
+		LatestVersion:   stepInfoPtr.LatestVersion,
+		GroupInfo:       stepInfoPtr.GroupInfo,
+		Step:            stepInfoPtr.Step,
+		DefinitionPth:   stepInfoPtr.DefinitionPth,
+	}
+
+	// Print step preparation errors before the step header box,
+	// other errors are printed within the step box.
+	if resultCode == models.StepRunStatusCodePreparationFailed && err != nil {
+		log.Errorf("Preparing Step (%s) failed: %s", pointers.StringWithDefault(stepInfoCopy.Step.Title, "missing title"), err)
+	}
+	bitrise.PrintRunningStepHeader(stepInfoPtr, step, stepIdxPtr)
+
+	errStr := ""
+	if err != nil {
+		errStr = err.Error()
+	}
+
+	stepResults := models.StepRunResultsModel{
+		StepInfo:   stepInfoCopy,
+		StepInputs: redactedStepInputs,
+		Status:     resultCode,
+		Idx:        buildRunResults.ResultsCount(),
+		RunTime:    time.Since(stepStartTime),
+		ErrorStr:   errStr,
+		ExitCode:   exitCode,
+		StartTime:  stepStartTime,
+	}
+
+	r.tracker.SendStepFinishedEvent(properties, analytics.StepResult{
+		Info:            prepareAnalyticsStepInfo(step, stepInfoPtr),
+		Status:          resultCode,
+		ErrorMessage:    errStr,
+		Timeout:         timeout,
+		NoOutputTimeout: noOutputTimeout,
+	})
+
+	switch resultCode {
+	case models.StepRunStatusCodeSuccess:
+		buildRunResults.SuccessSteps = append(buildRunResults.SuccessSteps, stepResults)
+	case models.StepRunStatusCodePreparationFailed:
+		buildRunResults.FailedSteps = append(buildRunResults.FailedSteps, stepResults)
+	case models.StepRunStatusCodeFailed:
+		log.Errorf("Step (%s) failed: %s", pointers.StringWithDefault(stepInfoCopy.Step.Title, "missing title"), err)
```

Comment:
This chande adds an additional error line (`ERRO:...`), is this intentional?

```

+------------------------------------------------------------------------------+
| (1) Fail wit exit code 2                                                     |
+------------------------------------------------------------------------------+
| id: script                                                                   |
| version:                                                                     |
| collection: https://github.com/bitrise-io/bitrise-steplib.git                |
| toolkit: bash                                                                |
| time: 2022-10-04T12:31:31+02:00                                              |
+------------------------------------------------------------------------------+
|                                                                              |
ERRO[12:31:31] Step (Fail wit exit code 2) failed: exit status 2 
|                                                                              |
+---+---------------------------------------------------------------+----------+
| x | Fail wit exit code 2 (exit code: 2)                           | 0.68 sec |
+---+---------------------------------------------------------------+----------+
| Issue tracker: https://github.com/bitrise-io/steps-script/issues             |
| Source: https://github.com/bitrise-io/steps-script                           |
+---+---------------------------------------------------------------+----------+
```

Reply:
Originally I didn't plan to change too many things other than the primary goal of the PR.
I left it accidentally as it is.

I can revert this change an make it work as it was working.
But it felt a bit strange how we print error messages:
- for StatusCodeFailed and StatusCodePreparationFailed we print only the non-exit status errors
- for StepRunStatusCodeFailedSkippable we print the error always (different error message for exit status error)

I thought always printing the error makes sense here and the client function should pass meaningful error messages.

Reply:
Original behavior restored in https://github.com/bitrise-io/bitrise/pull/816/commits/3dccb57e263e628d7287c7c2cc198b4978d7363c